### PR TITLE
[Docs] Add user guide for configuring authentication for Ray clusters

### PR DESCRIPTION
**Pull request by andrewsykim**
## Why are these changes needed?

Add a user guide for configuring authenticadtion and access control for Ray clusters using KubeRay and Kubernetes RBAC.

## Related issue number

Based on REP: https://github.com/ray-project/enhancements/blob/main/reps/2024-05-21-kuberay-authentication.md

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Moonquakes** [comment]
Hi @andrewsykim , I found that the protection in the document is mainly for submitting ray jobs through port 8265. Is there a corresponding protection mechanism and document for interactive access to Ray Cluster through port 10001? (It seems that the `ray.init` interface needs to be modified, but I am not sure if it is currently supported)

**Comment by andrewsykim** [comment]
@Moonquakes that's correct, for now we are starting with authentication for Ray dashboard only, in the future we will consider client port. @kevin85421 suggested we actually stay away from recommending interactive client but I can't recall the reasons

File: doc/source/cluster/kubernetes/user-guides/kuberay-auth.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+The output should be `yes`.
+
+## Submit a Ray Job with Authentication
+
+Now you can submit a Ray job using the service account's authentication token.
+
+Get a token for the `ray-user` service account and store it in the `RAY_JOB_HEADERS` environment variable:
+
+```bash
+export RAY_JOB_HEADERS="{\"Authorization\": \"Bearer $(kubectl create token ray-user)\"}"
```
Comment (kevin85421) [inline doc/source/cluster/kubernetes/user-guides/kuberay-auth.md:146]: Can you add a note that kubectl create token is only supported in kubectl versions >= 1.24 ([ref](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.24.md)) or update the "Prerequisites" section?

image

File: doc/source/cluster/kubernetes/user-guides/kuberay-auth.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
+(kuberay-auth)=
+
+# Configure Ray Clusters with Authentication and Access Control using KubeRay
+
+This guide demonstrates how to secure Ray clusters deployed with KubeRay by enabling authentication and access control using Kubernetes Role-Based Access Control (RBAC).
```
Comment (kevin85421) [inline doc/source/cluster/kubernetes/user-guides/kuberay-auth.md:5]: Can you add the current limitations to the documentation? For example, it currently doesn't support RayJob and RayService. How can I use the browser to view the dashboard? (follow up)

**Review by kevin85421 — COMMENTED** [review]
Nice!

**Comment by kevin85421** [comment]
@Moonquakes, yeah, the Ray maintainers currently recommend that users avoid using the Ray client. It has not been actively maintained for years due to some fundamental stability issues. We have been discussing Ray Client V2, but it is still in the planning stage for now.

File: doc/source/cluster/kubernetes/user-guides/kuberay-auth.md (hunk 1 above)
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/user-guides/kuberay-auth.md:146]: 1.24 has been EOL for a while now, it's probably safe to assume most users are using a newer version at this point

File: doc/source/cluster/kubernetes/user-guides/kuberay-auth.md (hunk 2 above)
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/user-guides/kuberay-auth.md:5]: I can add some instructions for accessing the dashboard via browser

Comment (andrewsykim) [reply doc/source/cluster/kubernetes/user-guides/kuberay-auth.md]: added instructions for using browser to access Ray dashboard

File: doc/source/cluster/kubernetes/user-guides/kuberay-auth.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+If you don't have a Kubernetes cluster, create one using the following command (or adapt it for your cloud provider):
+
+```bash
+gcloud container clusters create kuberay-cluster \
+    --num-nodes=2 --zone=us-west1-b --machine-type e2-standard-4
+```
+
+## Install the KubeRay Operator
+
```
[the remark this answers is not in the document]
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/user-guides/kuberay-auth.md:27]: done

File: doc/source/cluster/kubernetes/user-guides/kuberay-auth.md (hunk 1 above)
Comment (kevin85421) [reply doc/source/cluster/kubernetes/user-guides/kuberay-auth.md:141]: I followed this documentation and realized that I am using version 1.23.X on the client side. On the server side, I think most people use newer Kubernetes versions, but on the client side, users might not upgrade as frequently.

File: doc/source/cluster/kubernetes/user-guides.md [hunk 4]
```diff
@@ -27,6 +27,7 @@ user-guides/tls
 user-guides/k8s-autoscaler
 user-guides/static-ray-cluster-without-kuberay
 user-guides/kubectl-plugin
+user-guides/kuberay-auth
```
[the remark this answers is not in the document]
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/user-guides.md:30]: done

File: doc/source/cluster/kubernetes/user-guides/kuberay-auth.md (hunk 1 above)
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/user-guides/kuberay-auth.md:141]: good point, I'll add a note about this

Comment (andrewsykim) [reply doc/source/cluster/kubernetes/user-guides/kuberay-auth.md:141]: done

**Review by kevin85421 — APPROVED** [review]

**Comment by kevin85421** [comment]
I will ping our doc team to review this PR.

File: doc/source/cluster/kubernetes/user-guides/kuberay-auth.md [hunk 5]
```diff
@@ -0,0 +1,207 @@
+(kuberay-auth)=
+
+# Configure Ray Clusters with Authentication and Access Control using KubeRay
+
+This guide demonstrates how to secure Ray clusters deployed with KubeRay by enabling authentication and access control using Kubernetes Role-Based Access Control (RBAC). 
+
+> **Note:** This guide is only supported for the RayCluster custom resource.
```
[the remark this answers is not in the document]
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/user-guides/kuberay-auth.md:7]: I believe tokens last 1hour by default, you can specify the duration though:
```
kubectl create token ray-user --duration=24h
```

The server may return an expiration shorter than what the client requests though

[the remark this answers is not in the document]
Comment (kevin85421) [reply doc/source/cluster/kubernetes/user-guides/kuberay-auth.md:7]: Would you mind adding some comments regarding the expiration? Thanks!

[the remark this answers is not in the document]
Comment (andrewsykim) [reply doc/source/cluster/kubernetes/user-guides/kuberay-auth.md:7]: Updated the example to include --duration so users know the flag exists

**Comment by angelinalg** [comment]
Thank you for contributing to Ray docs. Just some style nits to be more consistent with our style guide. Thank you.

**Comment by andrewsykim** [comment]
@angelinalg I didn't see any comments for style nits, did you forget to submit the review?

**Comment by angelinalg** [comment]
So sorry for the oversight, @andrewsykim!

**Comment by Moonquakes** [comment]
> @Moonquakes, yeah, the Ray maintainers currently recommend that users avoid using the Ray client. It has not been actively maintained for years due to some fundamental stability issues. We have been discussing Ray Client V2, but it is still in the planning stage for now.

Thanks for your answer! However, Ray Client will be very helpful in the scenario of using Jupyter, which can greatly improve development efficiency. I look forward to the release of Ray Client v2!

**Comment by kevin85421** [comment]
> Ray Client will be very helpful in the scenario of using Jupyter, which can greatly improve development efficiency.

Agree. A good Ray client provides better UX comparing with `ray job submit`.

**Review by kevin85421 — APPROVED** [review]

**Pull request merged**